agent/engine/dockerauth/ecr.go

```diff
@@ -17,54 +17,115 @@ import (
 	"encoding/base64"
 	"fmt"
 	"strings"
+	"sync"
+	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/async"
 	"github.com/aws/amazon-ecs-agent/agent/ecr"
 	ecrapi "github.com/aws/amazon-ecs-agent/agent/ecr/model/ecr"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	"github.com/aws/aws-sdk-go/aws"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
 )
 
+type cacheKey struct {
+	region           string
+	roleARN          string
+	registryID       string
+	endpointOverride string
+}
+
 type ecrAuthProvider struct {
-	authData *api.ECRAuthData
-	client   ecr.ECRClient
+	tokenCache async.Cache
+	factory    ecr.ECRFactory
+	cacheLock  sync.RWMutex
 }
 
-const proxyEndpointScheme = "https://"
+const (
+	roundtripTimeout = 5 * time.Second
+	// MinimumJitterDuration is the minimum duration to mark the credentials
+	// as expired before it's actually expired
+	MinimumJitterDuration = 30 * time.Minute
+	proxyEndpointScheme   = "https://"
+)
+
+// String formats the cachKey as a string
+func (key *cacheKey) String() string {
+	return fmt.Sprintf("%s-%s-%s-%s", key.roleARN, key.region, key.registryID, key.endpointOverride)
+}
 
 // NewECRAuthProvider returns a DockerAuthProvider that can handle retrieve
 // credentials for pulling from Amazon EC2 Container Registry
-func NewECRAuthProvider(authData *api.ECRAuthData, clientFactory ecr.ECRFactory) DockerAuthProvider {
-	if authData == nil {
-		return &ecrAuthProvider{}
-	}
-	log.Tracef("Getting client in %s with endpoint %s", authData.Region, authData.EndpointOverride)
+func NewECRAuthProvider(ecrFactory ecr.ECRFactory, cache async.Cache) DockerAuthProvider {
 	return &ecrAuthProvider{
-		authData: authData,
-		client:   clientFactory.GetClient(authData.Region, authData.EndpointOverride),
+		tokenCache: cache,
+		factory:    ecrFactory,
 	}
 }
 
 // GetAuthconfig retrieves the correct auth configuration for the given repository
-func (authProvider *ecrAuthProvider) GetAuthconfig(image string) (docker.AuthConfiguration, error) {
-	if authProvider.authData == nil {
-		return docker.AuthConfiguration{}, fmt.Errorf("ecrAuthProvider cannot be used without AuthData")
+func (authProvider *ecrAuthProvider) GetAuthconfig(image string,
+	authData *api.ECRAuthData) (docker.AuthConfiguration, error) {
+	if authData == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("ecr auth: miissing container auth data")
+	}
+
+	// First try to get the token from cache, if the token does not exist,
+	// then call ECR api to get the new token
+	key := cacheKey{
+		region:           authData.Region,
+		endpointOverride: authData.EndpointOverride,
+		registryID:       authData.RegistryID,
+	}
+
+	// If the container is using execution role credentials to pull,
+	// add the roleARN as part of the cache key so that docker auth for
+	// containers pull with the same role can be cached
+	if !utils.ZeroOrNil(authData.GetPullCredentials()) {
```

Comment:
the value returned by `authData.GetPullCredentials()` is not pointer, it won't work with `!=nil`. The reason not return a pointer is to avoid data race.

Reply:
@samuelkarp are we trying to phase out the use of `utils.ZeroOrNil` or did you mean its not necessary in this case?

Reply:
Both.  It's reflection-based, which makes it not very performant (though the performance impact is likely negligible here), but my real objection to it is that it's [complicated and not super easy to understand](https://github.com/aws/amazon-ecs-agent/blob/f94beb472760dac30e6c1fccdcdc5e0a2e66cb60/agent/utils/utils.go#L39-L59).  It's really useful when we're dealing with operations over arbitrary structs (like in [`config.Merge`](https://github.com/aws/amazon-ecs-agent/blob/f94beb472760dac30e6c1fccdcdc5e0a2e66cb60/agent/config/config.go#L85-L100)), but I'd like to avoid it whenever we actually know the type of the thing we're comparing.